File: contests/lc/3583.CountSpecialTriplets/solution.py
from collections import defaultdict
from typing import List

MOD = 1_000_000_007


def count_special_triplets(nums: List[int]) -> int:
    """
    Count triplets (i, j, k) with i < j < k and nums[i] = nums[j]*2 = nums[k].
    Uses prefix/suffix frequency maps to tally matches in O(n).
    """
    suffix = defaultdict(int)
    for v in nums:
        suffix[v] += 1

    prefix = defaultdict(int)
    ans = 0

    for v in nums:
        suffix[v] -= 1
        target = v * 2
        ans = (ans + prefix[target] * suffix[target]) % MOD
        prefix[v] += 1

    return ans


class Solution:
    # LeetCode-style method name for compatibility.
    def specialTriplets(self, nums: List[int]) -> int:
        dic = {}
        dic2 = {}
        res = 0
        for num in nums:
            if num % 2 == 0 and num // 2 in dic2:
                res += dic2[num // 2]
                res %= 10**9 + 7
            if num * 2 in dic:
                if num in dic2:
                    dic2[num] += dic[num * 2]
                else:
                    dic2[num] = dic[num * 2]
            if num in dic:
                dic[num] += 1
            else:
                dic[num] = 1
        return res

    # Convenience wrapper that delegates to the shared implementation above.
    def count_special_triplets(self, nums: List[int]) -> int:
        return count_special_triplets(nums)


if __name__ == "__main__":
    solution = Solution()
    samples = [
        [6, 3, 6],
        [0, 1, 0, 0],
        [8, 4, 2, 8, 4],
    ]
    for nums in samples:
        print(f"nums={nums} -> {solution.count_special_triplets(nums)}")
        print(f"nums={nums} -> {solution.specialTriplets(nums)}")

File: contests/lc/3583.CountSpecialTriplets/solution.go
package main

import "fmt"

const mod int64 = 1_000_000_007

// countSpecialTriplets returns the number of index triplets (i, j, k)
// such that i < j < k and nums[i] = nums[j]*2 and nums[k] = nums[j]*2.
// It uses prefix/suffix frequency maps to count matches in O(n).
func countSpecialTriplets(nums []int) int {
	suffix := make(map[int]int64, len(nums))
	for _, v := range nums {
		suffix[v]++
	}

	prefix := make(map[int]int64, len(nums))
	var ans int64

	for _, v := range nums {
		suffix[v]--
		target := v * 2

		left := prefix[target]
		right := suffix[target]
		ans = (ans + left*right) % mod

		prefix[v]++
	}

	return int(ans)
}

func specialTriplets(nums []int) (ans int) {
	const mod int = 1e9 + 7

	seen := [200_002]int{}
	seen2 := [200_002]int{}

	for _, num := range nums {
		ans += seen2[num]
		seen2[2*num] += seen[2*num]
		seen[num]++
	}

	ans = (ans%mod + mod) % mod
	return
}

func main() {
	samples := [][]int{
		{6, 3, 6},
		{0, 1, 0, 0},
		{8, 4, 2, 8, 4},
	}

	for _, nums := range samples {
		fmt.Printf("nums=%v -> %d\n", nums, countSpecialTriplets(nums))
		fmt.Printf("nums=%v -> %d\n", nums, specialTriplets(nums))
	}
}
